common4j/src/main/com/microsoft/identity/common/java/authorities/Authority.java

```diff
@@ -157,9 +157,7 @@ public static Authority getAuthorityFromAuthorityUrl(String authorityUrl) {
             } else if (CIAM.equalsIgnoreCase(authorityTypeStr)) {
                 authority = new CIAMAuthority(authorityUrl);
             } else if (AAD_NA.equalsIgnoreCase(authorityTypeStr) && configuredAuthority instanceof NativeAuthCIAMAuthority) {
-                authority = new NativeAuthCIAMAuthority(
-                        authorityUrl,
-                        ((NativeAuthCIAMAuthority) configuredAuthority).getClientId());
+                authority = new CIAMAuthority(authorityUrl);
```

Comment:
Why do we want to return a CIAMAuthority in the case of native auth, rather than a NativeAuthAuthority? Don't our flows need a NativeAuthAuthority to complete the API flows?

Reply:
The CIAMAuthority will be transformed into NativeAuthAuthority in later stage using the current client id. `((NativeAuthCIAMAuthority) configuredAuthority).getClientId());` this line is the reason that cased persisted data in different SDK instances.

Reply:
Maybe I haven't understood the problem correctly. What exactly is the SDK behavior we want to achieve. Do we want to create a new authority object and add it to the known authorities list for each SDK instance? What happens if two instances use the same client id are we going to create two different objects and keep them in the list in this case?

Reply:
The SDK behavior we want to achieve is to create a new authority object and add it to the known authorities list for each different authority instance (the current SDK standard is that different authority url). Every time adding the authority, SDK clears up the authority list

```kotlin
PublicClientApplication.createNativeAuthApplication

config.getAuthorities().clear();
final Authority authorityObject = Authority.getAuthorityFromAuthorityUrl(authority, clientId);
authorityObject.setDefault(true);
config.getAuthorities().add(authorityObject);
```
And
```kotlin
NativeAuthPublicClientApplicationConfiguration.validateConfiguration

if (defaultAuthority is CIAMAuthority) {
    val nativeAuthAuthority = NativeAuthCIAMAuthority(
        authorityUrl = defaultAuthority.authorityUri.toString(),
        clientId = clientId
    )
   authorities.clear()
   authorities.add(nativeAuthAuthority)
}
```
So each SDK instance will only have one authority in its authority list.

For each E2E test file, such as SignUpEmailPasswordTest and SignUpEmailOTPTest, their SDK instance (application) are different. Each instance has its own configuration file and one authority url as well as one client id. One SDK instance has multiple authorities in the list in my mind only happends there are multiple authority urls in the configuration files, whic is out of native auth's scopes.

Reply:
@Yuki-YuXin  does `config.getAuthorities().clear();` clear the `knownAuthorities` list inside `Authority`?

Reply:
Then I don't think your reply to Giannis is correct. What we're trying to achieve is:
1. not change the fundamental MSAL code that retrieves the authority from `knownAuthorities`: we need to understand why this logic is here.
2. not change the fundamental MSAL code that determines whether an authority is known: it currently does that based on authority URL, and doesn't take into account the client ID for matching. Same as above; we need to understand this logic better, so we're not touching it for now.
3. the native auth flows can continue to look up authorities from the `knownAuthorities` list, but we can't rely on the client ID from previous instances. We're therefore going to change the way `Authority.getAuthorityFromAuthorityUrl` works: an authority URL should be passed as well as a client ID (optional generally, but expected for native auth), so that this method can return a fully formed `NativeAuthCIAMAuthority` with updated data.

Reply:
In terms of the first and second point, let me ask MSAL side offline for better explanation and understanding.